File: js/BeginnersSelection/09_Daydream.js
// https://atcoder.jp/contests/abs/tasks/arc065_a

Object.prototype.let = function(f) {
  return f(this.valueOf());
}

require('fs')
  .readFileSync('/dev/stdin', 'utf8')
  .trim()
  .let(s =>
    ['eraser', 'erase', 'dreamer', 'dream']
      .reduce(
        (arr, word) =>
          [].concat(...arr.map(x => x.split(word).filter(x => x))),
        [s]
      )
      .length == 0 ? 'YES' : 'NO'
  )
  .let(console.log);

// https://atcoder.jp/contests/abs/submissions/21477635

File: rb/BeginnersSelection/09_Daydream.rb
# https://atcoder.jp/contests/abs/tasks/arc065_a

class Object
  def let(f)
    f.call self
  end
end

gets
  .chomp
  .let(->(s) {
    ["eraser", "erase", "dreamer", "dream"]
      .reduce([s]) { |acc, word|
        acc.flat_map { |x|
          x.split(word).filter { |x| x != "" }
        }
      }
  })
  .size
  .let(->(x) { x == 0 ? "YES" : "NO" })
  .let(method(:puts))

# https://atcoder.jp/contests/abs/submissions/21525265
